pkg/querier/worker/frontend_processor.go

```diff
@@ -59,19 +65,24 @@ func (fp *frontendProcessor) notifyShutdown(ctx context.Context, conn *grpc.Clie
 }
 
 // runOne loops, trying to establish a stream to the frontend to begin request processing.
-func (fp *frontendProcessor) processQueriesOnSingleStream(ctx context.Context, conn *grpc.ClientConn, address string) {
-	client := frontendv1pb.NewFrontendClient(conn)
+func (fp *frontendProcessor) processQueriesOnSingleStream(workerCtx context.Context, conn *grpc.ClientConn, address string) {
+	client := fp.frontendClientFactory(conn)
+
+	// Run the querier loop (and so all the queries) in a dedicated context that we call the "execution context".
+	// The execution context is cancelled once the workerCtx is cancelled AND there's no inflight query executing.
+	exec := newExecutionContext(workerCtx, fp.log)
+	defer exec.cancel()
 
-	backoff := backoff.New(ctx, processorBackoffConfig)
+	backoff := backoff.New(exec.context(), processorBackoffConfig)
 	for backoff.Ongoing() {
-		c, err := client.Process(ctx)
+		c, err := client.Process(exec.context())
 		if err != nil {
 			level.Error(fp.log).Log("msg", "error contacting frontend", "address", address, "err", err)
 			backoff.Wait()
 			continue
 		}
 
-		if err := fp.process(c); err != nil {
+		if err := fp.process(c, exec); err != nil {
```

Comment:
Nit: One benefit of `newExecutionContext` returning `ctx, cancelFn, *atomic.Bool` is that we could only pass the bool to `process` function here. (Eg. as `queryInProgress`)

Reply:
Done!